File: September Challenge/Week 1/Java/Day_6/ImageOverlap.java
// Problem : https://leetcode.com/problems/image-overlap/
// Solution:

class Solution {
    public int largestOverlap(int[][] A, int[][] B) {
        int xStart = 0 - A.length;
        int yStart = 0 - A.length;
        xStart++;
        yStart++;
        int max=Integer.MIN_VALUE;
        for(int i = xStart; i<A.length;i++){
            for(int j = yStart; j<A.length;j++){
                max=Math.max(max,getOverlap(A,B,i,j));
            }
        }
        return max;
    }
    
    private int getOverlap(int[][] A, int[][] B, int X, int Y){
        int count=0;
        int idash=0;
        int jdash=0;
        for(int i=0;i<A.length;i++){
            for(int j=0;j<A.length;j++){
                idash = i-X;
                jdash = j-Y;
                if(idash<0 || idash>=A.length || jdash<0 || jdash>=A.length) 
                    continue;
                if((A[i][j]&B[idash][jdash])==1){
                    count++;
                }
            }
        }    
        return count;
    }
}

File: September Challenge/Week 1/CPP/Day_6/ImageOverlap.cpp
class Solution {
public:

int helper(vector<vector<int>> &A,vector<vector<int>> &B_padded,int xshift,int yshift)
{
    int x=0;
    
    for(int i=0;i<A.size();i++)
    {
        for(int j=0;j<A[0].size();j++)
        {
            x+=A[i][j]*B_padded[i+yshift][j+xshift];
        }
    }
    
    return x;
    
}

int largestOverlap(vector<vector<int>>& A, vector<vector<int>>& B) {
    
    
    int n=A.size();
    
    vector<vector<int>> B_padded(3*n-2,vector<int>(3*n-2,0));
    
    for(int i=0;i<n;i++)
    {
        for(int j=0;j<n;j++)
        {
            B_padded[i+n-1][j+n-1]=B[i][j];
        }
    }
    
    int result=0;
    
    for(int xshift=0;xshift<2*n-1;xshift++)
    {
        for(int yshift=0;yshift<2*n-1;yshift++)
        {
            result=max(result,helper(A,B_padded,xshift,yshift));
        }
    }
    
    return result;
    
}
};
